newspaper/mthreading.py

```diff
@@ -9,50 +9,58 @@
 __license__ = 'MIT'
 __copyright__ = 'Copyright 2014, Lucas Ou-Yang'
 
-import queue
-import traceback
-from threading import Thread
+from concurrent.futures import ThreadPoolExecutor, wait
+from threading import Lock
 
 from .configuration import Configuration
 
-
-class Worker(Thread):
-    """
-    Thread executing tasks from a given tasks queue.
-    """
-    def __init__(self, tasks, timeout_seconds):
-        Thread.__init__(self)
-        self.tasks = tasks
-        self.timeout = timeout_seconds
-        self.daemon = True
-        self.start()
-
-    def run(self):
-        while True:
-            try:
-                func, args, kargs = self.tasks.get(timeout=self.timeout)
-            except queue.Empty:
-                # Extra thread allocated, no job, exit gracefully
-                break
-            try:
-                func(*args, **kargs)
-            except Exception:
-                traceback.print_exc()
-
-            self.tasks.task_done()
-
+def perform_download_articles(paper):
+    return paper.download_articles
 
 class ThreadPool:
-    def __init__(self, num_threads, timeout_seconds):
-        self.tasks = queue.Queue(num_threads)
-        for _ in range(num_threads):
-            Worker(self.tasks, timeout_seconds)
+    def __init__(self, num_threads, timeout_seconds=0):
+        self.mutex = Lock()
+        self.executor = ThreadPoolExecutor(max_workers=num_threads)
+        self.timeout_seconds = timeout_seconds
+        self.futures = {}
 
     def add_task(self, func, *args, **kargs):
-        self.tasks.put((func, args, kargs))
+        """
+        Adds the given task, and returns a future for the completion.
+        """
+        f = self.executor.submit(func, *args, **kargs)
+
+        f.add_done_callback(self._remove_future)
+
+        self.mutex.acquire()
```

Comment:
```python
with self.mutex:
    self.futures[f] = 1
```
?

Reply:
Sorry, didn't know you could do that. Much better ...